### ISSUE TITLE
Training time increasing after 8-9k steps

### ISSUE THREAD

**Issue by sunilmtma** [trianglesplatting/triangle-splatting#3]
When i train traingle splatting for a bicycle scene, the training really slow down after 8-9k iterations as you can see here. I did not really changed anything in code. The 3dgs training time for this scene was 6 minutes for 30k iterations.

Is it normal or something else?

(triangle-splatting) schoudhary@SChoudhary:/mnt/c/Users/schoudhary/CESI/DEV/TS/triangle-splatting$ python train.py  -s /mnt/c/Users/schoudhary/CESI/DEV/Datasets/Dataset_IECON/Bicycle  -m /mnt/c/Users/schoudhary/CESI/DEV/Datasets/trainglesplatting/bicycle  --eval
Optimizing /mnt/c/Users/schoudhary/CESI/DEV/Datasets/trainglesplatting/bicycle
Setting up [LPIPS] perceptual loss: trunk [vgg], v[0.1], spatial [off]
/home/schoudhary/anaconda3/envs/triangle-splatting/lib/python3.11/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/schoudhary/anaconda3/envs/triangle-splatting/lib/python3.11/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=VGG16_Weights.IMAGENET1K_V1`. You can also use `weights=VGG16_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
Loading model from: /home/schoudhary/anaconda3/envs/triangle-splatting/lib/python3.11/site-packages/lpips/weights/v0.1/vgg.pth
Output folder: /mnt/c/Users/schoudhary/CESI/DEV/Datasets/trainglesplatting/bicycle [04/06 20:37:25]
Reading camera 360/360 [04/06 20:37:33]
Loading Training Cameras [04/06 20:37:33]
[ INFO ] Encountered quite large input images (>1.6K pixels width), rescaling to 1.6K.
 If this is not desired, please explicitly specify '--resolution/-r' as 1 [04/06 20:37:33]
Loading Test Cameras [04/06 20:38:31]
Training progress:  23%|██████████████████████████▏                                                                                     | 7000/30000 [06:08<41:11,  9.31it/s, Loss=0.00347][ITER 7000] Evaluating test: L1 0.0028620677534490825 PSNR 34.575957489013675 SSIM 0.9914652254846361 LPIPS 0.011088916038473447 [04/06 20:44:54]

[ITER 7000] Evaluating train: L1 0.0027066226350143554 PSNR 34.96193237304688 SSIM 0.9916416764259339 LPIPS 0.010863303020596505 [04/06 20:44:55]

[ITER 7000] Saving Triangles [04/06 20:44:55]
Training progress:  30%|████████████████████████████████▉                                                                            | 9060/30000 [19:31<58:27:29, 10.05s/it, Loss=0.01426]

**Comment by heldJan** [comment]
Hey @sunilmtma, add the following to your command line -i images_4, i.e. set python train.py -s /mnt/c/Users/schoudhary/CESI/DEV/Datasets/Dataset_IECON/Bicycle -m /mnt/c/Users/schoudhary/CESI/DEV/Datasets/trainglesplatting/bicycle --eval -i images_4

With this, the training should finish in about 35 minutes for the bicycle scene. Let me know if it works.

**Comment by sunilmtma** [comment]
Hi @heldJan , Thanks for your response and I am really sorry for the confusion. This bicycle dataset is not from SOTA papers but an synthetic dataset I created. So I was wondering, how can training really got slow down after 9k steps?

**Comment by heldJan** [comment]
Oh okay, I see. Maybe try setting the maximum number of triangles to a lower value, by default, it's fixed at 4M. You can change it using --max_shapes XXX.

**Comment by sunilmtma** [comment]
I see the issue now,

I was using images without background and by defalut it should take black background (as mentioned in train.py) but when I render test images it shows this with white background. **I'm surprised**. Due to this there were white splats (or triangles), were disturbing the quality and also after 7k steps the mesh file was too messy (can't really see the object).

So now I fixed it to white background and training time is closer to 15 minutes for 30k steps.

**Comment by sunilmtma** [comment]
Thank you for this information. Finally I can see triangles with good rendering quality.

![Image](https://github.com/user-attachments/assets/754cf0c3-e040-4a45-a1b7-0ed061cdaf11)

**Comment by heldJan** [comment]
Great!

**Comment by joevento** [comment]
I could observe a similar slow down with my real scene. From 0 to 7k iterations I was getting around 20-25 it/s (doing some other things at the same time so slight fluctuations) on a 4090 + xeon w5-3435x. After the evaluation and save the rest of the iterations were running noticeably slower at around 10 it/s. I didn't use any fancy arguments just: input output --eval. 

Looking at system utilization while running it again, it looks like its still running on all 32 threads after the 7k point but only one is doing anything. The others are sitting at 0.0% util, but the processes exist. Gpu is pinned at 100%.

Same thing happens on another dataset I created.

**Comment by heldJan** [comment]
Hi @joevento , maybe try logging the number of triangles during training. It could be that a large number of triangles is added after 7k iterations, which results in slower rendering.
The default number of triangles is 4M. You can use --max_shapes XXX to set a lower amount. Let me know if that works!

**Comment by tksharpless** [comment]
I had the same issue with my first test run, on one of the stock datasets mentioned in the paper, namely 'flowers'.  After 7000 iterations at 8.8 iter/sec it saved triangles, then jumped to about 3 sec/iter, and at 5 hours had increased to 25 sec/iter, with predicted completion 154 hours.  
I killed that, and retried with '-i images_4 --max_shapes 500000'.  This started at 27 it/sec, which had dropped to 17 it/sec by the 7000 mark.  It then continued at about 17 it/sec, with GPU compute nearly saturated but using only 4.3 of 8 GB vram (the first run had saturated vram) and less than one core's worth of CPU.  It maintained a stable predicted completion time of around 30 minutes until it actually did so, reporting
`[ITER 30000] Evaluating test: L1 0.06871211156249046 PSNR 19.981979630210184 SSIM 0.5109919119964946 LPIPS 0.3998662829399109 [29/07 15:45:44]

[ITER 30000] Evaluating train: L1 0.04892001301050186 PSNR 22.17103271484375 SSIM 0.6618358850479127 LPIPS 0.31941629052162174 [29/07 15:45:45]`
I'm satisfied that it is practical to run triangle_splatting under Windows on the Lenovo gaming laptop I use for developing photography software.  I will be mainly trying adapt it to solve simpler triangle optimization problems than these splats, so the achieved performance is OK.

**Comment by heldJan** [comment]
Hi @tksharpless , for outdoor scenes of MipNeRF360, you should use the -i images_4 flag, and for indoor scenes, -i images_2. 

I am glad that you already solved the issue.

Thanks.

**Comment by tksharpless** [comment]
@heldJan, would it be any good to use fewer images rather than lower resolution ones?  I'm working on ways to make good 3D models from sets of 50 or so photos.
